rust/src/transformations/polars/make_agg_trans/mod.rs

```diff
@@ -0,0 +1,232 @@
+use crate::combinators::assert_components_match;
+use num::Zero;
+use opendp_derive::bootstrap;
+use polars::prelude::*;
+
+use crate::core::{
+    Function, Metric, MetricSpace, StabilityMap, Transformation,
+};
+use crate::domains::{ExprDomain, LazyFrameDomain, LazyGroupByContext, LazyGroupByDomain, ExprMetric};
+use crate::error::*;
+use crate::traits::TotalOrd;
+
+#[bootstrap(ffi = false)]
+pub fn make_agg_trans<T: AggTransformation>(
+    input_domain: LazyGroupByDomain,
+    input_metric: T::InputMetric,
+    transformations: Vec<T>,
+) -> Fallible<Transformation<LazyGroupByDomain, LazyFrameDomain, T::InputMetric, <T::OutputMetric as ExprMetric<LazyGroupByContext>>::InnerMetric>>
+where
+    T::OutputMetric: ExprMetric<LazyGroupByContext>,
+    <T::OutputMetric as Metric>::Distance: TotalOrd + Zero,
+    (LazyGroupByDomain, T::InputMetric): MetricSpace,
+    (LazyFrameDomain, <T::OutputMetric as ExprMetric<LazyGroupByContext>>::InnerMetric): MetricSpace,
+{
+    // resolve transformations
+    let expr_input_domain = ExprDomain {
+        lazy_frame_domain: input_domain.lazy_frame_domain.clone(),
+        context: LazyGroupByContext {
+            columns: input_domain.grouping_columns.clone(),
+        },
+        active_column: None,
+    };
+
+    let transformations = (transformations.into_iter())
+        .map(|t| t.fix(&expr_input_domain, &input_metric))
+        .collect::<Fallible<Vec<_>>>()?;
+
+    let functions: Vec<_> = (transformations.iter())
+        .map(|t| t.function.clone())
+        .collect();
+
+    let function = Function::new_fallible(move |lazy_group: &LazyGroupBy| -> Fallible<LazyFrame> {
+        let exprs = (functions.iter())
+            .map(|t| t.eval(&(lazy_group.clone(), all())))
+            .map(|res| Ok(res?.1))
+            .collect::<Fallible<Vec<Expr>>>()?;
+
+        Ok(lazy_group.clone().agg(&exprs))
+    });
+
+    // output metric
+    let output_metric = (transformations.first())
+        .map(|t| &t.output_metric)
+        .ok_or_else(|| err!(MakeTransformation, "transformation list cannot be empty"))?
+        .clone();
+
+    transformations.iter().try_for_each(|t| {
+        Ok(assert_components_match!(
+            MetricMismatch,
+            t.output_metric,
+            output_metric
+        ))
+    })?;
+
+    // stability map
+    let stability_maps: Vec<_> = (transformations.iter())
+        .map(|t| t.stability_map.clone())
+        .collect();
+
+    let stability_map = StabilityMap::new_fallible(move |d_in| {
+        (stability_maps.iter())
+            .try_fold(<T::OutputMetric as Metric>::Distance::zero(), |acc, map| {
+                acc.total_max(map.eval(d_in)?)
+            })
+    });
```

Comment:
This is suspicious. I think the total max works here when we have one column, or when the transformations are all row-by-row over datasets. The proof may be achievable if we restricted the number of exprs to just one, when the metric is not a dataset metric? More consideration needed.